### PR TITLE
PADV 327 - Add steps to install extra private requirements

### PR DESCRIPTION
**Pull request by sergivalero20**
## Ticket

https://agile-jira.pearson.com/browse/PADV-327

## Description

This PR adds two new step to the workflow that helps to install extra private requirements and customized themes.

## Type of Change

- Adding two new steps to the workflow.

## Proof of Concept

This is a PoC, so the important to see or evaluate in this PR is each step in the workflow.

## How to test

- There is a branch called `vue/PADV-317` in openedx-platform that just contains one workflow. That workflow uses this Composite Action.
- There is a repository (only for test) created in Docker Hub called `pearsonopenedxops/tutor-test`.
- The workflow is activated when it push a commit to the `vue/PADV-317` branch.
- It is recommend to activate the workflow making a minor change on the code, for example change the tag of the image.

## Screenshot

The screenshot below shows the result of the last execution of the workflow.

image

## Reviewers

- [x] @anfbermudezme
- [x] @Jacatove
- [x] @Squirrel18
- [x]  @alexjmpb

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -29,6 +29,17 @@ inputs:
   docker-token:
     description: 'Docker Hub token.'
     required: true
+  extra-requirements:
+    description: 'Private repository to be cloned.'
+    required: false
+    default: 'false'
```
Comment (Jacatove) [inline action.yml:35]: shouldn't it be `false`?. @sergivalero20 
```suggestion
    default: false
```

File: action.yml [hunk 2]
```diff
@@ -49,6 +60,26 @@ runs:
...
+        repositories=("pearson-core")
+        for repository in "${repositories[@]}"; do
+          git clone https://${{ inputs.access-token }}@github.com/Pearson-Advance/$repository.git \
+            "$(tutor config printroot)/env/build/openedx/requirements/$repository"
+          echo "-e ./$repository" >> "$(tutor config printroot)/env/build/openedx/requirements/private.txt"
+          echo "Cloned "$repository" repository successfully."
+        done
+      shell: bash
+
+    - if: ${{ inputs.theme-repository != 'false' }}
```
Comment (Jacatove) [inline action.yml:76]: Maybe you can change the default to `""`,  the conditional becomes:
```suggestion
    - if: ${{ inputs.theme-repository != '' }}
```
@sergivalero20

File: action.yml [hunk 3]
```diff
@@ -49,6 +60,26 @@ runs:
       run: pip install tutor==${{ inputs.tutor-version }}
       shell: bash
 
+    - if: ${{ inputs.extra-requirements == 'true' }}
+      name: Install extra requirements
+      run: |
+        # repositores is an array that contains private repositories.
+        repositories=("pearson-core")
```
[the remark this answers is not in the document]
Comment (Squirrel18) [reply action.yml:67]: This is just a POC, no need to add more packages. @Jacatove 
You need to get the list of private repositories from an environment variable: https://docs.github.com/en/actions/learn-github-actions/contexts#vars-context @sergivalero20

File: action.yml [hunk 4]
```diff
@@ -49,6 +60,26 @@ runs:
...
+        for repository in "${repositories[@]}"; do
+          git clone https://${{ inputs.access-token }}@github.com/Pearson-Advance/$repository.git \
```
Comment (Squirrel18) [inline action.yml:69]: We need to add the capability of cloning the projects from a specific branch. @sergivalero20

**Review by Squirrel18 — CHANGES_REQUESTED** [review]

File: action.yml (hunk 1 above)
Comment (sergivalero20) [reply action.yml:35]: I tried to change it before, but I got this error: _**A string representing the default value. The default value is used when an input parameter isn't specified in a workflow file.**_. However I can change the default value to ''. @Jacatove

File: action.yml (hunk 4 above)
Comment (sergivalero20) [reply action.yml:69]: Ok, got it.

File: action.yml [hunk 5]
```diff
@@ -49,6 +60,26 @@ runs:
...
+          echo "-e ./$repository" >> "$(tutor config printroot)/env/build/openedx/requirements/private.txt"
+          echo "Cloned "$repository" repository successfully."
+        done
+      shell: bash
+
+    - if: ${{ inputs.theme-repository != 'false' }}
+      name: Install themes
+      run: |
+        git clone https://github.com/${{ inputs.theme-repository }}.git \
+          "$(tutor config printroot)/env/build/openedx/themes/myopenedxtheme"
```
[the remark this answers is not in the document]
Comment (sergivalero20) [reply action.yml:80]: I reviewed it and `open-release/juniper.stage` branch is the default branch @Jacatove 

image

**Comment by sergivalero20** [comment]
@Squirrel18 @Jacatove I added better comments in the README file and in each input of the Composite Action.

File: action.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Squirrel18) [reply action.yml:89]: We need to be able to change the branch, so it should be an additional parameter or attached to the themes repository URL. @sergivalero20

File: action.yml [hunk 6]
```diff
@@ -29,6 +29,23 @@ inputs:
   docker-token:
     description: 'Docker Hub token.'
     required: true
+  extra-requirements:
```
[the remark this answers is not in the document]
Comment (sergivalero20) [reply action.yml]: The last commit already contains that change. @Squirrel18 thanks.

File: README.md [hunk 7]
```diff
@@ -32,19 +32,26 @@ jobs:
           image-name: 'tutor-test'
           docker-username: ${{ secrets.DOCKERHUB_USERNAME }}
           docker-token: ${{ secrets.DOCKERHUB_TOKEN }}
+          extra-requirements: true
+          access-token: ${{ secrets.GH_PAT }}
```
Comment (anfbermudezme) [inline README.md:36]: Could you please change the name to github-token to make it clearer?

File: README.md [hunk 8]
```diff
@@ -32,19 +32,26 @@ jobs:
...
+| theme-repository   | String   |      No      | Repository that contains customized themes. You have to put it in this way: <owner>/<repository> (e.g., Pearson-Advance/openedx-themes)                                                                                                                                                                                                                                                                                                                                                |
+| access-token       | String   |      Yes     | Personal Access Token (PAT). If you have not configured it yet, enter in the follow link to see how create a PAT: https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token. When it is already configured, you can add it as a secrets for the repository. To see how create a secret enter in the next link: https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository. |
```
Comment (anfbermudezme) [inline README.md:55]: Same comment for change the name to github-token.

**Review by anfbermudezme — CHANGES_REQUESTED** [review]

File: README.md (hunk 7 above)
Comment (sergivalero20) [reply README.md:36]: Yes, sure!
Thanks @anfbermudezme

File: action.yml [hunk 9]
```diff
@@ -29,6 +29,27 @@ inputs:
   docker-token:
     description: 'Docker Hub token.'
     required: true
+  extra-private-requirements:
+    description: 'Bool input that indicates if It is necessary to install extra private requirements.'
+    required: false
+    default: 'false'
+  theme-repository:
```
Comment (Squirrel18) [inline action.yml:36]: This is a public repository, so I think we can include the full Github URL path. @sergivalero20

**Review by Squirrel18 — CHANGES_REQUESTED** [review]

File: action.yml (hunk 9 above)
Comment (sergivalero20) [reply action.yml:36]: Sure!
Change applied @Squirrel18

**Review by alexjmpb — APPROVED** [review]

**Review by anfbermudezme — APPROVED** [review]

**Review by Jacatove — APPROVED** [review]

**Review by Squirrel18 — APPROVED** [review]

**Review by anfbermudezme — APPROVED** [review]

**Pull request merged**